### ISSUE TITLE
Bug: QWEN2 MoE imatrix contains nan's after generating it

### ISSUE THREAD

**Issue by legraphista** [ggerganov/llama.cpp#7816]
### What happened?

Steps:
1. Generate an F32 of https://huggingface.co/Qwen/Qwen2-57B-A14B-Instruct with this patch applied https://huggingface.co/Qwen/Qwen2-57B-A14B-Instruct/discussions/2 (for quickness, I generated Q8 for imatrix calc and F32 for quants)
3. create an imatrix offloading some layers to CUDA using this text file https://huggingface.co/legraphista/Qwen2-57B-A14B-Instruct-IMat-GGUF/blob/main/imatrix.dataset
4. attempt to apply imatrix for quantization

### Name and Version

```
version: 3105 (7027b27d)
built with cc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0 for x86_64-linux-gnu
```

### What operating system are you seeing the problem on?

Linux

### Relevant log output

```shell
$ llama.cpp/imatrix -m Qwen2-57B-A14B-Instruct-GGUF/Qwen2-57B-A14B-Instruct.Q8_0.gguf -f imatrix.dataset -ngl 10 -c 512 -b 512 --threads 32 -o imatrix.dat
llama_model_loader: loaded meta data with 23 key-value pairs and 479 tensors from Qwen2-57B-A14B-Instruct-GGUF/Qwen2-57B-A14B-Instruct.Q8_0.gguf.hardlink.gguf (version GGUF V3 (latest))
llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
llama_model_loader: - kv   0:                       general.architecture str              = qwen2moe
llama_model_loader: - kv   1:                               general.name str              = Qwen2-57B-A14B-Instruct
llama_model_loader: - kv   2:                       qwen2moe.block_count u32              = 28
llama_model_loader: - kv   3:                    qwen2moe.context_length u32              = 32768
llama_model_loader: - kv   4:                  qwen2moe.embedding_length u32              = 3584
llama_model_loader: - kv   5:               qwen2moe.feed_forward_length u32              = 20480
llama_model_loader: - kv   6:              qwen2moe.attention.head_count u32              = 28
llama_model_loader: - kv   7:           qwen2moe.attention.head_count_kv u32              = 4
llama_model_loader: - kv   8:                    qwen2moe.rope.freq_base f32              = 1000000.000000
llama_model_loader: - kv   9:  qwen2moe.attention.layer_norm_rms_epsilon f32              = 0.000001
llama_model_loader: - kv  10:                 qwen2moe.expert_used_count u32              = 8
llama_model_loader: - kv  11:                          general.file_type u32              = 7
llama_model_loader: - kv  12:                      qwen2moe.expert_count u32              = 64
llama_model_loader: - kv  13:                       tokenizer.ggml.model str              = gpt2
llama_model_loader: - kv  14:                         tokenizer.ggml.pre str              = qwen2
llama_model_loader: - kv  15:                      tokenizer.ggml.tokens arr[str,151936]  = ["!", "\"", "#", "$", "%", "&", "'", ...
llama_model_loader: - kv  16:                  tokenizer.ggml.token_type arr[i32,151936]  = [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...
llama_model_loader: - kv  17:                      tokenizer.ggml.merges arr[str,151387]  = ["Ġ Ġ", "ĠĠ ĠĠ", "i n", "Ġ t",...
llama_model_loader: - kv  18:                tokenizer.ggml.eos_token_id u32              = 151645
llama_model_loader: - kv  19:            tokenizer.ggml.padding_token_id u32              = 151643
llama_model_loader: - kv  20:                tokenizer.ggml.bos_token_id u32              = 151643
llama_model_loader: - kv  21:                    tokenizer.chat_template str              = {% for message in messages %}{% if lo...
llama_model_loader: - kv  22:               general.quantization_version u32              = 2
llama_model_loader: - type  f32:  197 tensors
llama_model_loader: - type q8_0:  282 tensors
llm_load_vocab: special tokens cache size = 293
llm_load_vocab: token to piece cache size = 0.9338 MB
llm_load_print_meta: format           = GGUF V3 (latest)
llm_load_print_meta: arch             = qwen2moe
llm_load_print_meta: vocab type       = BPE
llm_load_print_meta: n_vocab          = 151936
llm_load_print_meta: n_merges         = 151387
llm_load_print_meta: n_ctx_train      = 32768
llm_load_print_meta: n_embd           = 3584
llm_load_print_meta: n_head           = 28
llm_load_print_meta: n_head_kv        = 4
llm_load_print_meta: n_layer          = 28
llm_load_print_meta: n_rot            = 128
llm_load_print_meta: n_embd_head_k    = 128
llm_load_print_meta: n_embd_head_v    = 128
llm_load_print_meta: n_gqa            = 7
llm_load_print_meta: n_embd_k_gqa     = 512
llm_load_print_meta: n_embd_v_gqa     = 512
llm_load_print_meta: f_norm_eps       = 0.0e+00
llm_load_print_meta: f_norm_rms_eps   = 1.0e-06
llm_load_print_meta: f_clamp_kqv      = 0.0e+00
llm_load_print_meta: f_max_alibi_bias = 0.0e+00
llm_load_print_meta: f_logit_scale    = 0.0e+00
llm_load_print_meta: n_ff             = 20480
llm_load_print_meta: n_expert         = 64
llm_load_print_meta: n_expert_used    = 8
llm_load_print_meta: causal attn      = 1
llm_load_print_meta: pooling type     = 0
llm_load_print_meta: rope type        = 2
llm_load_print_meta: rope scaling     = linear
llm_load_print_meta: freq_base_train  = 1000000.0
llm_load_print_meta: freq_scale_train = 1
llm_load_print_meta: n_ctx_orig_yarn  = 32768
llm_load_print_meta: rope_finetuned   = unknown
llm_load_print_meta: ssm_d_conv       = 0
llm_load_print_meta: ssm_d_inner      = 0
llm_load_print_meta: ssm_d_state      = 0
llm_load_print_meta: ssm_dt_rank      = 0
llm_load_print_meta: model type       = ?B
llm_load_print_meta: model ftype      = Q8_0
llm_load_print_meta: model params     = 57.41 B
llm_load_print_meta: model size       = 56.83 GiB (8.50 BPW) 
llm_load_print_meta: general.name     = Qwen2-57B-A14B-Instruct
llm_load_print_meta: BOS token        = 151643 '<|endoftext|>'
llm_load_print_meta: EOS token        = 151645 '<|im_end|>'
llm_load_print_meta: PAD token        = 151643 '<|endoftext|>'
llm_load_print_meta: LF token         = 148848 'ÄĬ'
llm_load_print_meta: EOT token        = 151645 '<|im_end|>'
ggml_cuda_init: GGML_CUDA_FORCE_MMQ:   no
ggml_cuda_init: CUDA_USE_TENSOR_CORES: yes
ggml_cuda_init: found 1 CUDA devices:
  Device 0: NVIDIA GeForce RTX 4090, compute capability 8.9, VMM: yes
llm_load_tensors: ggml ctx size =    0.43 MiB
llm_load_tensors: offloading 10 repeating layers to GPU
llm_load_tensors: offloaded 10/29 layers to GPU
llm_load_tensors:        CPU buffer size = 58190.19 MiB
llm_load_tensors:      CUDA0 buffer size = 20388.09 MiB
.............................................................................................
llama_new_context_with_model: n_ctx      = 512
llama_new_context_with_model: n_batch    = 512
llama_new_context_with_model: n_ubatch   = 512
llama_new_context_with_model: flash_attn = 0
llama_new_context_with_model: freq_base  = 1000000.0
llama_new_context_with_model: freq_scale = 1
llama_kv_cache_init:  CUDA_Host KV buffer size =    18.00 MiB
llama_kv_cache_init:      CUDA0 KV buffer size =    10.00 MiB
llama_new_context_with_model: KV self size  =   28.00 MiB, K (f16):   14.00 MiB, V (f16):   14.00 MiB
llama_new_context_with_model:  CUDA_Host  output buffer size =     0.58 MiB
llama_new_context_with_model:      CUDA0 compute buffer size =   862.52 MiB
llama_new_context_with_model:  CUDA_Host compute buffer size =     8.01 MiB
llama_new_context_with_model: graph nodes  = 1910
llama_new_context_with_model: graph splits = 346

system_info: n_threads = 32 / 32 | AVX = 1 | AVX_VNNI = 0 | AVX2 = 1 | AVX512 = 1 | AVX512_VBMI = 1 | AVX512_VNNI = 1 | AVX512_BF16 = 1 | FMA = 1 | NEON = 0 | SVE = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | SSSE3 = 1 | VSX = 0 | MATMUL_INT8 = 0 | LLAMAFILE = 1 | 
compute_imatrix: tokenizing the input ..
compute_imatrix: tokenization took 135.885 ms
compute_imatrix: computing over 128 chunks with batch_size 512
compute_imatrix: 5.16 seconds per pass - ETA 11.00 minutes
[1]4.9575,[2]3.4740,[3]3.1781,[4]3.5055,[5]3.4223,[6]3.1119,[7]3.2700,[8]3.2929,[9]3.7110,
save_imatrix: stored collected data after 10 chunks in imatrix.dat
[10]3.6668,[11]3.6189,[12]3.9895,[13]4.4287,[14]4.6725,[15]5.0404,[16]5.3116,[17]5.4825,[18]5.6501,[19]5.4929,
save_imatrix: stored collected data after 20 chunks in imatrix.dat
[20]5.6098,[21]5.6066,[22]5.6538,[23]5.6008,[24]5.8175,[25]5.9804,[26]5.8777,[27]5.9822,[28]6.0379,[29]6.1864,
save_imatrix: stored collected data after 30 chunks in imatrix.dat
[30]6.1506,[31]5.9725,[32]5.7566,[33]5.6509,[34]5.5608,[35]5.5154,[36]5.5215,[37]5.5629,[38]5.6044,[39]5.5722,
save_imatrix: stored collected data after 40 chunks in imatrix.dat
[40]5.6990,[41]5.7423,[42]5.9444,[43]6.1088,[44]6.2754,[45]6.4090,[46]6.5149,[47]6.4206,[48]6.4557,[49]6.5417,
save_imatrix: stored collected data after 50 chunks in imatrix.dat
[50]6.6096,[51]6.5266,[52]6.5961,[53]6.7232,[54]6.8086,[55]6.8662,[56]6.9126,[57]6.9567,[58]6.9946,[59]7.0255,
save_imatrix: stored collected data after 60 chunks in imatrix.dat
[60]7.0625,[61]7.0266,[62]6.9920,[63]7.0473,[64]7.1158,[65]7.0680,[66]7.0670,[67]7.0863,[68]7.0229,[69]6.9689,
save_imatrix: stored collected data after 70 chunks in imatrix.dat
[70]6.9623,[71]6.9463,[72]6.9431,[73]6.9588,[74]6.9001,[75]6.8524,[76]6.8061,[77]6.7868,[78]6.7740,[79]6.7511,
save_imatrix: stored collected data after 80 chunks in imatrix.dat
[80]6.6806,[81]6.7077,[82]6.7016,[83]6.6616,[84]6.6788,[85]6.6970,[86]6.6610,[87]6.6396,[88]6.6213,[89]6.6418,
save_imatrix: stored collected data after 90 chunks in imatrix.dat
[90]6.6657,[91]6.6567,[92]6.6193,[93]6.5897,[94]6.5406,[95]6.5056,[96]6.4673,[97]6.4356,[98]6.3961,[99]6.3680,
save_imatrix: stored collected data after 100 chunks in imatrix.dat
[100]6.3803,[101]6.4046,[102]6.4739,[103]6.5511,[104]6.6015,[105]6.7051,[106]6.7713,[107]6.8045,[108]6.7913,[109]6.7797,
save_imatrix: stored collected data after 110 chunks in imatrix.dat
[110]6.7746,[111]6.7326,[112]6.6627,[113]6.6131,[114]6.6655,[115]6.6699,[116]6.6776,[117]6.6962,[118]6.7280,[119]6.7310,
save_imatrix: stored collected data after 120 chunks in imatrix.dat
[120]6.7268,[121]6.7397,[122]6.7134,[123]6.7513,[124]6.7813,[125]6.8029,[126]6.8390,[127]6.8695,[128]6.9092,
save_imatrix: stored collected data after 128 chunks in imatrix.dat

llama_print_timings:        load time =    8351.52 ms
llama_print_timings:      sample time =       0.00 ms /     1 runs   (    0.00 ms per token,      inf tokens per second)
llama_print_timings: prompt eval time =  675786.55 ms / 65536 tokens (   10.31 ms per token,    96.98 tokens per second)
llama_print_timings:        eval time =       0.00 ms /     1 runs   (    0.00 ms per token,      inf tokens per second)
llama_print_timings:       total time =  680556.83 ms / 65537 tokens

Final estimate: PPL = 6.9092 +/- 0.09472
```

```
$ llama.cpp/quantize --imatrix imatrix.dat Qwen2-57B-A14B-Instruct-GGUF/Qwen2-57B-A14B-Instruct.gguf test.gguf Q2_K
load_imatrix: imatrix dataset='imatrix.dataset'
load_imatrix: loaded 336 importance matrix entries from imatrix.dat computed on 128 chunks
prepare_imatrix: have 336 importance matrix entries
main: build = 3105 (7027b27d)
main: built with cc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0 for x86_64-linux-gnu
main: quantizing 'Qwen2-57B-A14B-Instruct-GGUF/Qwen2-57B-A14B-Instruct.gguf' to 'test.gguf' as Q2_K
llama_model_loader: loaded meta data with 23 key-value pairs and 479 tensors from Qwen2-57B-A14B-Instruct-GGUF/Qwen2-57B-A14B-Instruct.gguf (version GGUF V3 (latest))
llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
llama_model_loader: - kv   0:                       general.architecture str              = qwen2moe
llama_model_loader: - kv   1:                               general.name str              = Qwen2-57B-A14B-Instruct
llama_model_loader: - kv   2:                       qwen2moe.block_count u32              = 28
llama_model_loader: - kv   3:                    qwen2moe.context_length u32              = 32768
llama_model_loader: - kv   4:                  qwen2moe.embedding_length u32              = 3584
llama_model_loader: - kv   5:               qwen2moe.feed_forward_length u32              = 20480
llama_model_loader: - kv   6:              qwen2moe.attention.head_count u32              = 28
llama_model_loader: - kv   7:           qwen2moe.attention.head_count_kv u32              = 4
llama_model_loader: - kv   8:                    qwen2moe.rope.freq_base f32              = 1000000.000000
llama_model_loader: - kv   9:  qwen2moe.attention.layer_norm_rms_epsilon f32              = 0.000001
llama_model_loader: - kv  10:                 qwen2moe.expert_used_count u32              = 8
llama_model_loader: - kv  11:                          general.file_type u32              = 0
llama_model_loader: - kv  12:                      qwen2moe.expert_count u32              = 64
llama_model_loader: - kv  13:                       tokenizer.ggml.model str              = gpt2
llama_model_loader: - kv  14:                         tokenizer.ggml.pre str              = qwen2
llama_model_loader: - kv  15:                      tokenizer.ggml.tokens arr[str,151936]  = ["!", "\"", "#", "$", "%", "&", "'", ...
llama_model_loader: - kv  16:                  tokenizer.ggml.token_type arr[i32,151936]  = [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...
llama_model_loader: - kv  17:                      tokenizer.ggml.merges arr[str,151387]  = ["Ġ Ġ", "ĠĠ ĠĠ", "i n", "Ġ t",...
llama_model_loader: - kv  18:                tokenizer.ggml.eos_token_id u32              = 151645
llama_model_loader: - kv  19:            tokenizer.ggml.padding_token_id u32              = 151643
llama_model_loader: - kv  20:                tokenizer.ggml.bos_token_id u32              = 151643
llama_model_loader: - kv  21:                    tokenizer.chat_template str              = {% for message in messages %}{% if lo...
llama_model_loader: - kv  22:               general.quantization_version u32              = 2
llama_model_loader: - type  f32:  479 tensors
================================ Have weights data with 336 entries
llama_model_quantize: failed to quantize: imatrix contains non-finite value -nan

main: failed to quantize model from 'Qwen2-57B-A14B-Instruct-GGUF/Qwen2-57B-A14B-Instruct.gguf'
```

Possibly related to https://github.com/ggerganov/llama.cpp/issues/7805

**Comment by CISC** [comment]
I have this too, I think it's because of the following line:
https://github.com/ggerganov/llama.cpp/blob/c9ee7118d5644dd3df70ea6878b36a9761616aab/examples/imatrix/imatrix.cpp#L234
It looks like counts is 0 sometimes, but I don't understand why, @slaren any idea what is going on here?

**Comment by legraphista** [comment]
Just throwing an idea out there, could it be that the imatrix dataset text file is not varied enough to activate all experts?

I was using this one, and for example, it does not contain any Chinese : https://gist.githubusercontent.com/bartowski1182/eb213dccb3571f863da82e99418f81e8/raw/b2869d80f5c16fd7082594248e80144677736635/calibration_datav3.txt

**Comment by ggerganov** [comment]
#7833 should fix this problem

> Generate an F32 of [huggingface.co/Qwen/Qwen2-57B-A14B-Instruct](https://huggingface.co/Qwen/Qwen2-57B-A14B-Instruct) with this patch applied [huggingface.co/Qwen/Qwen2-57B-A14B-Instruct/discussions/2](https://huggingface.co/Qwen/Qwen2-57B-A14B-Instruct/discussions/2)

This patch is not correct - instead the `convert-hf-to-gguf.py` script should be updated to take into account the `moe_intermediate_size` and `shared_expert_intermediate_size` fields from `config.json`

**Comment by CISC** [comment]
> This patch is not correct - instead the `convert-hf-to-gguf.py` script should be updated to take into account the `moe_intermediate_size` and `shared_expert_intermediate_size` fields from `config.json`

Aha, can this explain the 0-weight issue?

**Comment by legraphista** [comment]
> #7833 should fix this problem
> 
> > Generate an F32 of [huggingface.co/Qwen/Qwen2-57B-A14B-Instruct](https://huggingface.co/Qwen/Qwen2-57B-A14B-Instruct) with this patch applied [huggingface.co/Qwen/Qwen2-57B-A14B-Instruct/discussions/2](https://huggingface.co/Qwen/Qwen2-57B-A14B-Instruct/discussions/2)
> 
> This patch is not correct - instead the `convert-hf-to-gguf.py` script should be updated to take into account the `moe_intermediate_size` and `shared_expert_intermediate_size` fields from `config.json`

Thanks @ggerganov for the pointer. I've made a pr here https://github.com/ggerganov/llama.cpp/pull/7835

**Comment by CISC** [comment]
It looks like Qwen2-57B-A14B is impossible to quantize with an imatrix, first of all it's impossible to fully activate all the tensors (I've tried massive multilingual diverse datasets with no luck), some barely reaching ~60%, which will lead to an incomplete imatrix, making lower i-quants fail.

I've debugged this extensively now, and the reason seems to be there are large chunks of abnormal weights (lots of 0.0, but also other subnormal values) which cause major issues with the imatrix calculations. I've tried fudging the imatrix and clamping the quantized weights, but it just produces garbage so far.

TLDR: I think Qwen2-57B-A14B is "broken" and needs fixing by Qwen.
